symbolic-base/src/ZkFold/Base/Data/Logical.hs

```diff
@@ -0,0 +1,216 @@
+
+or :: (Boolean b, Foldable t) => t b -> b
+or = any id
+
+class Boolean (BooleanOf a) => Conditional a where
+
+  type BooleanOf a :: Type
+  type BooleanOf a = GBooleanOf (G.Rep a)
+
+  bool :: a -> a -> BooleanOf a -> a
+  default bool ::
+    ( G.Generic a
+    , GConditional (G.Rep a)
+    , BooleanOf a ~ GBooleanOf (G.Rep a)
+    ) => a -> a -> BooleanOf a -> a
+  bool f t b = G.to (gbool (G.from f) (G.from t) b)
+
+instance
+  ( Conditional x0
+  , Conditional x1
+  , BooleanOf x0 ~ BooleanOf x1
+  ) => Conditional (x0,x1)
+
+instance
+  ( Conditional x0
+  , Conditional x1
+  , Conditional x2
+  , BooleanOf x0 ~ BooleanOf x1
+  , BooleanOf x1 ~ BooleanOf x2
+  ) => Conditional (x0,x1,x2)
+
+instance
+  ( Conditional x0
+  , Conditional x1
+  , Conditional x2
+  , Conditional x3
+  , BooleanOf x0 ~ BooleanOf x1
+  , BooleanOf x1 ~ BooleanOf x2
+  , BooleanOf x2 ~ BooleanOf x3
+  ) => Conditional (x0,x1,x2,x3)
+
+instance (R.Representable v, Conditional x)
+  => Conditional (G.Generically1 v x) where
+    type BooleanOf (G.Generically1 v x) = BooleanOf x
+    bool (G.Generically1 vf) (G.Generically1 vt) b =
+      G.Generically1 (R.mzipWithRep (\f t -> bool f t b) vf vt)
+
+deriving via G.Generically1 (Vector n) x
+  instance (KnownNat n, Conditional x) => Conditional (Vector n x)
+
+deriving via G.Generically1 ((->) x) y
+  instance Conditional y => Conditional (x -> y)
+
+instance Conditional (Identity x) where
+  type BooleanOf (Identity x) = H.Bool
+  bool = H.bool
+
+ifThenElse :: Conditional a => BooleanOf a -> a -> a -> a
+ifThenElse b t f = bool f t b
+
+class Conditional a => Eq a where
+
+  (==) :: a -> a -> BooleanOf a
+  default (==) ::
+    ( G.Generic a
+    , GEq (G.Rep a)
+    , BooleanOf a ~ GBooleanOf (G.Rep a)
+    ) => a -> a -> BooleanOf a
+  x == y = geq (G.from x) (G.from y)
+
+  (/=) :: a -> a -> BooleanOf a
+  default (/=) ::
+    ( G.Generic a
+    , GEq (G.Rep a)
+    , BooleanOf a ~ GBooleanOf (G.Rep a)
+    ) => a -> a -> BooleanOf a
+  x /= y = gneq (G.from x) (G.from y)
+
+instance
+  ( Eq x0, Eq x1
+  , BooleanOf x0 ~ BooleanOf x1
+  ) => Eq (x0,x1)
+
+instance
+  ( Eq x0
+  , Eq x1
+  , Eq x2
+  , BooleanOf x0 ~ BooleanOf x1
+  , BooleanOf x1 ~ BooleanOf x2
+  ) => Eq (x0,x1,x2)
+
+instance
+  ( Eq x0
+  , Eq x1
+  , Eq x2
+  , Eq x3
+  , BooleanOf x0 ~ BooleanOf x1
+  , BooleanOf x1 ~ BooleanOf x2
+  , BooleanOf x2 ~ BooleanOf x3
+  ) => Eq (x0,x1,x2,x3)
+
+instance (Foldable v, R.Representable v, Eq x)
+  => Eq (G.Generically1 v x) where
+    G.Generically1 vx == G.Generically1 vy =
+      and (R.mzipWithRep (==) vx vy)
+    G.Generically1 vx /= G.Generically1 vy =
+      or (R.mzipWithRep (==) vx vy)
+
+deriving via G.Generically1 (Vector n) x
+  instance (KnownNat n, Eq x) => Eq (Vector n x)
+
+instance H.Eq x => Eq (Identity x) where
+  (==) = (H.==)
+  (/=) = (H./=)
+
+elem :: (Eq a, Foldable t) => a -> t a -> BooleanOf a
+elem x = any (== x)
+
+class Boolean (GBooleanOf f) => GConditional f where
```

Comment:
Suggestion: move deriving stuff to corresponding `.Generic` modules (e.g. `ZkFold.Base.Data.Eq.Generic`) to make modules shorter and more readable. I guess instances for `K1` would still need to be defined in the main module, but this still would look cleaner and understandable (because it would be near instances of the main class for tuples which need it)

Reply:
So 5 modules instead of 1? 🙃 I don't understand how this would work and not require mutually importing modules. I split it into 3 for now.

Reply:
ok, maybe at least visually separate the part on Generic deriving from the rest by a comment